imagebuildah/stage_executor.go

```diff
@@ -1521,6 +1574,124 @@ func (s *StageExecutor) tagExistingImage(ctx context.Context, cacheID, output st
 	return img.ID, ref, nil
 }
 
+// generateCacheKey returns a computed digest for the current STEP
+// running its history and diff against a hash algorithm and this
+// generated CacheKey is further used by buildah to lock and decide
+// tag for the intermeidate image which can be pushed and pulled to/from
+// the remote repository.
+func (s *StageExecutor) generateCacheKey(ctx context.Context, currNode *parser.Node, addedContentDigest string, buildAddsLayer bool) (string, error) {
+	h := sha256.New()
+	var baseHistory []v1.History
+	var manifestType string
+	var err error
+	if s.builder.FromImageID != "" {
+		manifestType, baseHistory, _, err = s.executor.getImageTypeAndHistoryAndDiffIDs(ctx, s.builder.FromImageID)
+		if err != nil {
+			return "", fmt.Errorf("error getting history of base image %q: %w", s.builder.FromImageID, err)
+		}
+	}
+	createdBy := s.getCreatedBy(currNode, addedContentDigest)
+	fmt.Fprintf(h, "%t\n", buildAddsLayer)
+	fmt.Fprintln(h, createdBy)
+	fmt.Fprintln(h, manifestType)
+	for _, element := range baseHistory {
+		fmt.Fprintln(h, element.CreatedBy)
+		fmt.Fprintln(h, element.Author)
+		fmt.Fprintln(h, element.Comment)
+		fmt.Fprintf(h, "%t\n", element.EmptyLayer)
+		fmt.Fprintln(h)
+	}
+	return fmt.Sprintf("%x", h.Sum(nil)), nil
+}
+
+// pushCache takes the image id of intermediate image and attempts
+// to perform push at the remote repository with cacheKey as the tag.
+// Returns error if fails otherwise returns nil.
+func (s *StageExecutor) pushCache(ctx context.Context, src, destSpec string) error {
+	logrus.Debugf("trying to push cache to dest: %+v from src:%+v", destSpec, src)
+	dest, err := alltransports.ParseImageName(destSpec)
+	// Add the docker:// transport to see if user neglected it while
+	// specifying --cache-to `<dest>`.
+	if err != nil {
+		destTransport := strings.Split(destSpec, ":")[0]
```

Comment:
I know that @mtrmac frowns on this.  I believe their should be containers/image libraries for this.

Reply:
Ah I see but following line is just here to check if valid transport was specified but we were still unable to ParseImageName in such case we don't attempt to add `docker://`, line below this split already uses `Get(` from `image/v5/transports` so I think we are already using `c/image` here but If there is even simpler way to check this then I'm ok to use that.

Reply:
- Conceptually text parsing like that belongs at the CLI level; APIs should deal with well-typed data. Compare `buildah.Push(… dest types.ImageReference)` vs. the heuristics living in `cmd/buildah`. Besides conceptual cleanliness, formatting errors should be discovered before any actual expensive build starts.
- If there must be a heuristic, I’d prefer them not to be copy&pasted; they are inevitably going to drift apart (as… they already did in the Buildah codebase, into about 6 variants at least, and I can’t see when one would use one over the other). E.g. if this is intentionally using the `cmd/buildah/push.go` heuristic, fine please split that into a function, name and document it well (that’s the hardest part!) and call that function from both places, instead of copy&pasting.
- BTW. note that the semantics/parsing of `pullCache` and `pushCache` are very different. Not that I know how to make them exactly the same, apart from mandating the `alltransports.ParseImageName` format. Ugh.

WRT the narrow c/image API question,
```go
 		if t := alltransports.TransportFromImageName(destSpec); t != nil { // << this
 			return err
 		}

 		if strings.Contains(destSpec, "://") {
 			return err
 		}

 		dest2, err2 := docker.ParseReference("//"+destSpec) // << this, not unnecessarily alltransports again
 		if err2 != nil {
 			return err
 		}
 		dest = dest2
 		logrus.Debugf("Assuming docker:// as the transport method for cache DESTINATION: %s", destSpec) // possibly update the output here
 	}

Reply:
Moved to CLI.